core/src/test/java/org/ehcache/EhcacheBasicPutEventsTests.java

```diff
@@ -0,0 +1,194 @@
+    this.store = spy(fakeStore);
+    final Ehcache<String, String> ehcache = this.getEhcache(null);
+    ehcache.put("key", "value");
+    verify(cacheEventListener, times(1)).onEvent(argThat(isCreated));
+    try {
+      ehcache.getRuntimeConfiguration().deregisterCacheEventListener(cacheEventListener);
+      fail();
+    } catch (UnsupportedOperationException e) {
+      //expected
+    }
+  }
+
+  @Test
+  public void testPutNoStoreEntryCacheAccessExceptionCacheLoaderWriterException() throws Exception {
+    final FakeStore fakeStore = new FakeStore(Collections.<String, String>emptyMap());
+    this.store = spy(fakeStore);
+    doThrow(new CacheAccessException("")).when(this.store).compute(eq("key"), getAnyBiFunction());
+    doThrow(new Exception()).when(this.cacheLoaderWriter).write("key", "value");
+    final Ehcache<String, String> ehcache = this.getEhcache(this.cacheLoaderWriter);
+    try {
+      ehcache.put("key", "value");
+      fail();
+    } catch (CacheWritingException e) {
+      // Expected
+    }
+    verify(cacheEventListener, never()).onEvent(Matchers.<CacheEvent<String, String>>any());
+    try {
+      ehcache.getRuntimeConfiguration().deregisterCacheEventListener(cacheEventListener);
+      fail();
+    } catch (UnsupportedOperationException e) {
+      //expected
+    }
+  }
+
+
+  @Test
+  public void testPutExceptionCacheLoaderWriterException() throws Exception {
+    final FakeStore fakeStore = new FakeStore(Collections.singletonMap("key", "oldValue"));
+    this.store = spy(fakeStore);
+    doThrow(new Exception()).when(this.cacheLoaderWriter).write("key", "value");
+    final Ehcache<String, String> ehcache = this.getEhcache(this.cacheLoaderWriter);
+    try {
+      ehcache.put("key", "value");
+      fail();
+    } catch (CacheWritingException e) {
+      // Expected
+    }
+    verify(cacheEventListener, never()).onEvent(Matchers.<CacheEvent<String, String>>any());
+    try {
+      ehcache.getRuntimeConfiguration().deregisterCacheEventListener(cacheEventListener);
+      fail();
+    } catch (UnsupportedOperationException e) {
+      //expected
+    }
+  }
+
+  @Test
+  public void testPutHasStoreEntryCacheAccessExceptionNoCacheLoaderWriter() throws Exception {
+    final FakeStore fakeStore = new FakeStore(Collections.singletonMap("key", "oldValue"));
+    this.store = spy(fakeStore);
+    doThrow(new CacheAccessException("")).when(this.store).compute(eq("key"), getAnyBiFunction());
+
+    final Ehcache<String, String> ehcache = this.getEhcache(null);
+
+    ehcache.put("key", "value");
+    verify(cacheEventListener, never()).onEvent(Matchers.<CacheEvent<String, String>>any());
+    try {
+      ehcache.getRuntimeConfiguration().deregisterCacheEventListener(cacheEventListener);
+      fail();
+    } catch (UnsupportedOperationException e) {
+      //expected
+    }
+  }
+
+  private Ehcache<String, String> getEhcache(final CacheLoaderWriter<String, String> cacheLoaderWriter) {
```

Comment:
this could be extracted to EhcacheBasicCrudBase in order to avoid code duplication